src/python/WMCore/MicroService/MSTransferor/MSTransferor.py

```diff
@@ -676,3 +730,115 @@ def _diskPNNs(self, pnnList):
             else:
                 diskPNNs.add(pnn)
         return diskPNNs
+
+    def updateSites(self, rec):
+        """
+        Update sites API provides asynchronous update of Site info.
+
+        :param rec: JSON payload with the following data structures:
+                    {'workflow': <wflow name>, 'SiteWhiteList' ['T1', ...], 'SiteBlackList': ['T2',...]}
+        :return: acknowledge dict to upstream caller (ReqMgr2)
+        """
+        # preserve provided payload to local file system
+        wflow = rec['workflow']
+        status = self.saveData(wflow, rec)
+        if status == 'ok':
+            return []
+        err = MSTransferorStorageError(rec)
+        self.logger.error(err)
+        return [err.error()]
+
+    def saveData(self, wflow, rec):
+        """
+        Save workflow data to persistent storage
+        :param wflow: name of workflow
+        :param rec: record to save
+        :return: status of this operation
+        """
+        try:
+            fname = '{}/{}'.format(self.storage, wflow)
+            with open(fname, 'w', encoding="utf-8") as ostream:
+                ostream.write(json.dumps(rec))
+            return 'ok'
+        except Exception as exp:
+            msg = "Unable to save workflow '%s' to storage=%s. Error: %s", wflow, self.storage, str(exp)
+            self.logger.exception(msg)
+            return str(exp)
+
+    def readData(self, wflow):
+        """
+        Read workflow data from persistent storage
+        :return: data or None
+        """
+        fname = '{}/{}'.format(self.storage, wflow)
+        data = None
+        with open(fname, 'r', encoding="utf-8") as istream:
+            data = json.load(istream)
+        return data
+
+    def _updateSites(self, wflow, rseList):
+        """
+        Internal update sites API to perform actual work (to be called from execute daemon):
+        - read persistent storage area
+        - for every found workflow area read its payload
+        - for every workflow perform the following set of steps:
+          1. accessing the workflow transfer document to find out the rule ids
+          2. if a site was added to the SiteBlacklist, that site needs to be removed from the rule ids RSE expression
+          3. if a site was added to the SiteWhitelist, that site needs to be added to the rule ids RSE expression
+          4. the new rule ids need to be persisted in the transfer document, replacing those rules that are no longer valid.
```

Comment:
I  think a 5th use case is missing here:

5. if a site has been removed from the workflow's SiteWhiteList

Which is supposed to trigger the same set of actions as 2, but still the init conditions differ. With a small correction - we  may  need to wait for anything that is already running at the site to finish. This is valid in for the workflows we update in `running-open` state. So to summarize in formal logic:
```
if 5. and wflow-status is not  running-open:
   repeat logic from 2.
else:
  check for running jobs at the site,
  if None:
    repeat logic from 2.
  else:
    do nothing until we have running jobs at the site completed.
```

Or something of the sort, no need the first guess of mine to be accepted as correct.

Reply:
this is no longer relevant as Alan requested to change logic and `_updateSites` is no longer required.